src/main/java/org/nrnb/gsoc/enrichment/ui/EnrichmentTableModel.java

```diff
@@ -260,6 +260,96 @@ private boolean inSource(List<TermSource> sources, String termName) {
         return false;
     }
 
+    public void removeRedundantTerms(List<TermSource> sources, double cutoff) {
+        List<CyRow> rows = cyTable.getAllRows();
+        Long[] rowArray = new Long[rows.size()];
+        int i = 0;
+        for (CyRow row : rows) {
+            String termCategory = row.get(EnrichmentTerm.colSource, String.class);
+            if (sources.size() == 0 || inCategory(sources, termCategory)) {
+                rowArray[i] = row.get(EnrichmentTerm.colID, Long.class);
+                i++;
+            }
+        }
+
+        rowNames = removeRedundancy(rowArray, i, cutoff);
+
+        fireTableDataChanged();
+    }
+
+    private boolean inCategory(List<TermSource> categories, String termName) {
+        for (TermSource tc: categories) {
+            if (tc.getName().equals(termName))
+                return true;
+        }
+        return false;
+    }
+
+    private Long[] removeRedundancy(Long[] rowArray, int length, double cutoff) {
+        // Sort by pValue
+        Long[] sortedArray = pValueSort(rowArray, length);
+
+        // Initialize with the most significant term
+        List<Long> currentTerms = new ArrayList<Long>();
+        currentTerms.add(sortedArray[0]);
+        for (int i = 1; i < length; i++) {
+            if (jaccard(currentTerms, sortedArray[i]) < cutoff)
+                currentTerms.add(sortedArray[i]);
+        }
+        return(currentTerms.toArray(new Long[1]));
+    }
+
+    private Long[] pValueSort(Long[] rowArray, int length) {
+        // Already sorted, I think...
```

Comment:
Add a todo in the comments to verify if it is sorted or not.

Reply:
Done